docs/fixing-warnings.md

```diff
@@ -0,0 +1,189 @@
+Let's take another look at the original example:
+
+```C#
+string s = Console.ReadLine();
+Type type = Type.GetType(s);
+foreach (var m in type.GetMethods())
+{
+    Console.WriteLine(m.Name);
+}
+```
+
+In the example above, the real problem is `Console.ReadLine()`. Because *any* type could
+be read the trimmer has no way to know if you need methods on `System.Tuple` or `System.Guid`
+or any other type. On the other hand, if your code looked like,
+
+```C#
+Type type = typeof(System.Tuple);
+foreach (var m in type.GetMethods())
+{
+    Console.WriteLine(m.Name);
+}
+```
+
+This would be fine. Here the trimmer can see the exact type being referenced: `System.Tuple`. Now
+it can use flow analysis to determine that it needs to keep all public methods. So where does
+`DynamicallyAccessMembers` come in? What happens if the reflection is split across methods?
+
+```C#
+void M1()
+{
+    M2(typeof(System.Tuple));
+}
+void M2(Type type)
+{
+    // Trim analysis warning IL2070: net6.Program.M2(Type): 'this' argument does not satisfy
+    // 'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'System.Type.GetMethods()'. The
+    // parameter 'type' of method 'net6.Program.M2(Type)' does not have matching annotations. The
+    // source value must declare at least the same requirements as those declared on the target
+    // location it is assigned to.
+    var methods = type.GetMethods();
+    ...
+}
+```
+
+Now you see a warning. For performance and stability flow analysis isn't performed between
+methods, so annotation is needed to pass information upward, from the reflection call
+(`GetMethods`) to the source of the `Type` (`typeof`). In the above example, the trimmer warning
+is stating that `GetMethods` requires the `PublicMethods` annotation on types, but the `type`
+variable doesn't have the same requirement. In other words, we need to pass the requirements from
+`GetMethods` up to the caller:
+
+```C#
+void M1()
+{
+    M2(typeof(System.Tuple));
+}
+void M2(
+    [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type type)
+{
+    var methods = type.GetMethods();
+    ...
+}
+```
+
+Now the warning disappears, because the trimmer knows exactly which members to preserve, and
+which type(s) to preserve them on. In general, this is the best way to deal with
+`DynamicallyAccessedMembers` warnings: add annotations so the trimmer knows what to preserve.
+
+As with `RequiresUnreferencedCode`, adding `RequiresUnreferencedCode` or
+`UnconditionalSuppressMessage` attributes also works, but none of these options make the code
+compatible with trimming, while adding `DynamicallyAccessedMembers` does.
+
+## Conclusion
+
+This description should cover the most common situations you end up in while trimming your
+application. Over time we'll continue to improve the diagnostic experience and provide more streamlined
+ways to make your code trim-compatible.
```

Comment:
> Over time we'll continue to improve the diagnostic experience and provide more streamlined
ways to make your code trim-compatible.

Is this claim correct?

Reply:
Yup! Better warning messages, analyzers for `DynamicallyAccessedMembers`, better perf, etc